mpqp/core/instruction/gates/controlled_gate.py

```diff
@@ -39,3 +42,116 @@ def __init__(
+
+        if nb_qubits != 0:
+            max_qubit = max(control, target) + 1
+            if nb_qubits < max_qubit:
+                raise ValueError(f"nb_qubits must be at least {max_qubit}")
+        else:
+            min_qubit = min(control, target)
+            control -= min_qubit
+            target -= min_qubit
+            nb_qubits = abs(control - target) + 1
+
+        zero = np.diag([1, 0]).astype(np.complex64)
+        one = np.diag([0, 1]).astype(np.complex64)
+        non_controlled_gate = self.non_controlled_gate.to_matrix()
+        I2 = np.eye(2, dtype=np.complex64)
+
+        control_matrix = zero if control == 0 else I2
+        target_matrix = (
+            one if control == 0 else (non_controlled_gate if target == 0 else I2)
+        )
+
+        for i in range(1, nb_qubits):
+            if i == control:
+                target_matrix = np.kron(target_matrix, one)
+                control_matrix = np.kron(control_matrix, zero)
+            elif i == target:
+                target_matrix = np.kron(target_matrix, non_controlled_gate)
+                control_matrix = np.kron(control_matrix, I2)
+            else:
+                target_matrix = np.kron(target_matrix, I2)
+                control_matrix = np.kron(control_matrix, I2)
+
+        return control_matrix + target_matrix
+
+    def _multi_control_gate_to_matrix(self, nb_qubits: int = 0) -> Matrix:
+        import numpy as np
+
+        from mpqp.core.instruction.gates.native_gates import SWAP
+
+        controls, targets = self.controls, self.targets
+        min_qubit, max_qubit = min(self.connections()), max(self.connections())
+
+        # If nb_qubits is not provided, calculate the necessary number of minimal qubits
+        if nb_qubits == 0:
+            nb_qubits = max_qubit - min_qubit + 1
+            controls = [x - min_qubit for x in controls]
+            targets = [x - min_qubit for x in targets]
+        elif nb_qubits < max_qubit + 1:
+            raise ValueError(f"nb_qubits must be at least {max_qubit + 1}")
+
+        # Get the canonical matrix and extend it to the correct size
+        canonical_matrix = self.to_canonical_matrix()
+        if canonical_matrix.shape[0] < 2**nb_qubits:
+            canonical_matrix = np.kron(
+                canonical_matrix, np.eye(2**nb_qubits // canonical_matrix.shape[0])
+            )
+
+        swaps = []
+        # qubit_types if a representation of the all qubits to follow target and control with swap
+        # assuming that canonical_matrix start with control and then target
+        qubit_types = {i: "None" for i in range(nb_qubits)}
+        qubit_types.update({i: "control" for i in range(len(controls))})
+        qubit_types.update({i + len(controls): "target" for i in range(len(targets))})
+
+        def swap_and_update(qubits: list[int], target_type: str):
+            for qubit in sorted(qubits):
+                if qubit_types[qubit] != target_type:
```

Comment:
actually, this works for controls but not for targets, so this shoud be
reworked, I'll try to see that tomorrow